src/app/AttributeCache.cpp

```diff
@@ -219,5 +219,75 @@ CHIP_ERROR AttributeCache::GetStatus(const ConcreteAttributePath & path, StatusI
     return CHIP_NO_ERROR;
 }
 
+struct compare
+{
+    bool operator()(const ConcreteDataAttributePathWithSize & x, const ConcreteDataAttributePathWithSize & y) const
+    {
+        return x.mSize < y.mSize;
+    }
+};
+
+uint32_t AttributeCache::OnUpdateDataVersionFilterList(DataVersionFilterIBs::Builder & aDataVersionFilterIBsBuilder,
+                                                       const Span<DataVersionFilter> & aDataVersionFilters)
+{
+    uint32_t number = 0;
+    CHIP_ERROR err  = CHIP_NO_ERROR;
+    TLV::TLVWriter backup;
+
+    std::set<ConcreteDataAttributePathWithSize, compare> clusterSet;
+    for (auto path = mChangedAttributeSet.begin(); path != mChangedAttributeSet.end(); path++)
+    {
+        ConcreteDataAttributePathWithSize cluster(*path);
+        UpdateClusterSize(cluster);
+        clusterSet.insert(cluster);
+    }
+
+    for (auto path = clusterSet.rbegin(); path != clusterSet.rend(); path++)
+    {
+        bool isIntersectDataVersionFilter = false;
+        DataVersionFilter dataVersionFilter(path->mEndpointId, path->mClusterId, path->mDataVersion);
+        aDataVersionFilterIBsBuilder.Checkpoint(backup);
+        VerifyOrExit(dataVersionFilter.IsValidDataVersionFilter(), err = CHIP_ERROR_INCORRECT_STATE);
+
+        // if the particular cached data version intersects with user provided data version filter, skip the cached one, and respect
+        // the user-provided one
+        for (auto & filter : aDataVersionFilters)
+        {
+            if (filter.IsSameCluster(*path))
+            {
+                isIntersectDataVersionFilter = true;
+                break;
+            }
+        }
+        if (isIntersectDataVersionFilter)
+        {
+            continue;
+        }
+
+        DataVersionFilterIB::Builder & filter = aDataVersionFilterIBsBuilder.CreateDataVersionFilter();
+        SuccessOrExit(err = aDataVersionFilterIBsBuilder.GetError());
+        ClusterPathIB::Builder & filterPath = filter.CreatePath();
+        SuccessOrExit(err = filter.GetError());
+        SuccessOrExit(err = filterPath.Endpoint(dataVersionFilter.mEndpointId)
+                                .Cluster(dataVersionFilter.mClusterId)
+                                .EndOfClusterPathIB()
+                                .GetError());
+        SuccessOrExit(err = filter.DataVersion(dataVersionFilter.mDataVersion.Value()).EndOfDataVersionFilterIB().GetError());
+        ChipLogProgress(
+            DataManagement, "Update DataVersionFilter: Endpoint=%" PRIu16 " Cluster=" ChipLogFormatMEI " Version=%" PRIu32,
+            dataVersionFilter.mEndpointId, ChipLogValueMEI(dataVersionFilter.mClusterId), dataVersionFilter.mDataVersion.Value());
+
+        number++;
+    }
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogProgress(DataManagement, "OnUpdateDataVersionFilterList rollbacks");
+        aDataVersionFilterIBsBuilder.Rollback(backup);
```

Comment:
I'd rather the cache (and every other potential implementor of this function) not be responsible for rollback. This should happen in just once place in the IM, key'ed off of a CHIP_OUT_OF_MEMORY error.

Reply:
To ensure you get the check-pointing right, you'll need to update the `DataVersionFilterIB::Builder` to do the checkpoint automatically when you call `CreateDataVersionFilter`. That then becomes much nicer as well in that you don't have to explicitly checkpoint in the code that uses that class.

Reply:
it seems I need make OnUpdateDataVersionFilterList to accept either ReadRequestMessage or SubscribeRequestIB, then request.EncodeDataVersionFilter(filter->first.mEndpointId, filter->first.mClusterId, filter->first.mDataVersion.Value()); inside request, it would automatically rollback when no memory, it seems I have to use same OnUpdateDataVersionFilterList to handle read and subscribe at the same time since they are different encoding class.
Need to think about it a while, creating the follow-up it on it.

Reply:
https://github.com/project-chip/connectedhomeip/issues/17034